### PR TITLE
`unnecessary_owned_empty_strings`

### PR DESCRIPTION
**Pull request by yoav-lavi**
[`unnecessary_owned_empty_strings`]

Fixes https://github.com/rust-lang/rust-clippy/issues/8650

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

[lint_naming]: https://rust-lang.github.io/rfcs/0344-conventions-galore.html#lints

changelog: Adds `unnecessary_owned_empty_strings`, a lint that detects passing owned empty strings to a function expecting `&str`

### CODE REVIEW

**Comment by yoav-lavi** [comment]
I need to do an additional check which does not currently run due to pulling from source and `cargo` being missing from the latest nightly, will try again tomorrow

**Comment by yoav-lavi** [comment]
Seems like this currently does not catch `"12345".split(&String::new())`, I've update the example in the meantime but needs to be checked why that specific example doesn't work

**Comment by Jarcho** [comment]
I wouldn't try to handle `str::split` right away. It takes a generic type bound by `Pattern` which both `str` and `String` implement.
It's doable, but a lot more involved.

**Comment by yoav-lavi** [comment]
Got it, thanks for the tip @Jarcho!

**Comment by yoav-lavi** [comment]
The discussion on Reddit brought up that you may want to use `&String::new()` for type consistency or due to the fact that  `&str` is a fat pointer while `&String` isn't, but I'm not sure that distinction is relevant when discussing `""` vs `&String::new()`.

Making sure that these aren't issues that should prevent this lint from being added

**Comment by Jarcho** [comment]
Irrelevant in this case. This lint is for any time a `&str` is expected and `&String::new()` is given. Since `&str` is expected a fat pointer will be created anyways.

**Comment by mikerite** [comment]
Tested on a bunch of crates. 3 warnings in yaml-rust. Fix worked.

```
---> yaml-rust-0.4.5/src/scanner.rs:700:60                            
warning: detected a usage of `&String::new()` for a function expecting a `&str` argument
   --> src/scanner.rs:700:60
    |
700 |         let prefix = self.scan_tag_uri(true, is_secondary, &String::new(), mark)?;
    |                                                            ^^^^^^^^^^^^^^ help: try: `""`
    |
    = note: requested on the command line with `-W clippy::unnecessary-string-new`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_string_new


---> yaml-rust-0.4.5/src/scanner.rs:736:54                            
warning: detected a usage of `&String::new()` for a function expecting a `&str` argument
   --> src/scanner.rs:736:54
    |
736 |             suffix = self.scan_tag_uri(false, false, &String::new(), &start_mark)?;
    |                                                      ^^^^^^^^^^^^^^ help: try: `""`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_string_new


---> yaml-rust-0.4.5/src/scanner.rs:754:62                            
warning: detected a usage of `&String::new()` for a function expecting a `&str` argument
   --> src/scanner.rs:754:62
    |
754 |                 suffix = self.scan_tag_uri(false, secondary, &String::new(), &start_mark)?;
    |                                                              ^^^^^^^^^^^^^^ help: try: `""`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_string_new
```

File: clippy_lints/src/unnecessary_string_new.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            if let Some(fn_def_id) = cx.qpath_res(qpath, fun.hir_id).opt_def_id();
+            if match_def_path(cx, fn_def_id, &paths::STRING_NEW); //  ["alloc", "string", "String", "new"]
+            if let ty::Ref(_, inner_str, _) = cx.typeck_results().expr_ty_adjusted(expr).kind();
+            if inner_str.is_str();
+            then {
+                span_lint_and_sugg(
+                    cx,
+                    UNNECESSARY_STRING_NEW,
+                    expr.span,
+                    "detected a usage of `&String::new()` for a function expecting a `&str` argument",
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_string_new.rs:50]:
```suggestion
                    "usage of `&String::new()` for a function expecting a `&str` argument",
```

File: clippy_lints/src/unnecessary_string_new.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryStringNew {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        let applicability = Applicability::MachineApplicable;
+        if_chain! {
+            if let ExprKind::AddrOf(.., inner_expr) = expr.kind;
+            if let ExprKind::Call(fun, _) = inner_expr.kind;
+            if let ExprKind::Path(ref qpath) = fun.kind;
+            if let Some(fn_def_id) = cx.qpath_res(qpath, fun.hir_id).opt_def_id();
+            if match_def_path(cx, fn_def_id, &paths::STRING_NEW); //  ["alloc", "string", "String", "new"]
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_string_new.rs:42]: I think a quite easy extension that could be added in this PR is to also lint `String::from("")`. See [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ec7975456f076477153bcb51eccbaf59)

File: clippy_lints/src/unnecessary_string_new.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    style,
+    "detects cases of `&String::new()` being passed as an argument to a function expecting `&str`"
+}
+declare_lint_pass!(UnnecessaryStringNew => [UNNECESSARY_STRING_NEW]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryStringNew {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        let applicability = Applicability::MachineApplicable;
+        if_chain! {
+            if let ExprKind::AddrOf(.., inner_expr) = expr.kind;
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_string_new.rs:38]: As mentioned on Zulip, I would also check for the mutability here and that it is a `Ref`:
```suggestion
            if let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Not, inner_expr) = expr.kind;
```

**Review by flip1995 — CHANGES_REQUESTED** [review]

File: clippy_lints/src/unnecessary_string_new.rs (hunk 2 above)
Comment (yoav-lavi) [reply clippy_lints/src/unnecessary_string_new.rs:42]: Will take a look at this, although I'm wondering if there's ever a case where you'd want to use `String::from("")` over `String::new()`, that might be it's own lint. WDYT?

Comment (yoav-lavi) [reply clippy_lints/src/unnecessary_string_new.rs]: Added

File: rust-toolchain [hunk 4]
```diff
@@ -1,3 +1,3 @@
 [toolchain]
-channel = "nightly-2022-04-07"
+channel = "nightly-2022-04-08"
```
Comment (yoav-lavi) [inline rust-toolchain:2]: Didn't mean to commit this, reverting.

Comment (yoav-lavi) [reply rust-toolchain]: Reverted

File: clippy_lints/src/unnecessary_owned_empty_string.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    #[clippy::version = "1.61.0"]
+    pub UNNECESSARY_OWNED_EMPTY_STRING,
+    style,
+    "detects cases of references to owned empty strings being passed as an argument to a function expecting `&str`"
+}
+declare_lint_pass!(UnnecessaryOwnedEmptyString => [UNNECESSARY_OWNED_EMPTY_STRING]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryOwnedEmptyString {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        let applicability = Applicability::MachineApplicable;
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_owned_empty_string.rs:38]: Just write `Applicability::MachineApplicable` directly in the `span_lint` function.

**Review by flip1995 — COMMENTED** [review]
Thanks, impl LGTM, just 2 more NITs

File: clippy_lints/src/unnecessary_owned_empty_string.rs (hunk 5 above)
Comment (yoav-lavi) [reply clippy_lints/src/unnecessary_owned_empty_string.rs]: Done

File: clippy_lints/src/unnecessary_owned_empty_string.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    ///
+    /// ### Example
+    /// ```rust
+    /// vec!["1", "2", "3"].join(&String::new());
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// vec!["1", "2", "3"].join("");
+    /// ```
+    #[clippy::version = "1.61.0"]
```
[the remark this answers is not in the document]
Comment (yoav-lavi) [reply clippy_lints/src/unnecessary_owned_empty_string.rs]: Done

**Review by flip1995 — APPROVED** [review]
Thanks, please squash your commits, so we can merge this :rocket:

**Comment by flip1995** [comment]
@bors r+

Thanks!

File: clippy_lints/src/unnecessary_owned_empty_string.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    /// ### Example
+    /// ```rust
+    /// vec!["1", "2", "3"].join(&String::new());
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// vec!["1", "2", "3"].join("");
+    /// ```
+    #[clippy::version = "1.62.0"]
+    pub UNNECESSARY_OWNED_EMPTY_STRING,
```
Comment (flip1995) [inline clippy_lints/src/unnecessary_owned_empty_string.rs:30]:
```suggestion
    pub UNNECESSARY_OWNED_EMPTY_STRINGS,
```

Lint names are usually better in the plural form, when they aren't describing a specific function name. This usually reads better in combination with `allow`.

This change will require `cargo dev update_lints`.

Comment (yoav-lavi) [reply clippy_lints/src/unnecessary_owned_empty_string.rs]: Done

**Pull request merged**